### PR TITLE
Universe setup tests

### PR DESCRIPTION
**Pull request by norbertsj**

### CODE REVIEW

File: src/protocol/warper/manager/errors.ts
Comment (alan-parsiq) [inline src/protocol/warper/manager/errors.ts]: I might be picky, but I would have two proposals:
1. either we could have /src/protocol/warperp/warper-manager/errors.ts (justifying that by referencing /contracts/warper/ structure)
2. or since, this formality might not be needed - to just to /src/protocol/warper/errors.ts and there would WARPER_MANAGER_CUSTOM_ERRORS or WARPERS_CUSTOM_ERRORS, so that all related to warper domain contracts/libraries inside one file

I am not really sure for now which certainly to go. (1) is more formal (and matches /contracts directory structure) and more overhead, (2) is less formal, but less overhead.

Comment (alan-parsiq) [reply src/protocol/warper/manager/errors.ts]: Would be also cool to follow the convention for WARPER_MANAGER_ERRORS -> WARPER_MANAGER_CUSTOM_ERRORS

Comment (alan-parsiq) [reply src/protocol/warper/manager/errors.ts]: Also, check the errors here - they are not all from the WarperManager. 

I can justify why we had this way previously - because before moving to hardhat-chai-matchers we had waffle-matchers and they were mostly like .revertsWith(warperManager, ...), so passing contract (the same conversation we had earlier today :) ).

Check from those which are WarperManager, IWarperManager or Warpers - and move either to separate file in separate directory (pattern No1) or in the same but IWARPER_MANAGER_CUSTOM_ERRORS, WARPERS_CUSTOM_ERRORS (No2) - actually this one looks more favorable, as less directories and overhead.

But from other perspective less formal!

File: src/protocol/warper/preset-factory/errors.ts
Comment (alan-parsiq) [inline src/protocol/warper/preset-factory/errors.ts]: WARPER_PRESET_CUSTOM_ERRORS -> IWARPER_PRESET_FACTORY_CUSTOM_ERRORS

I am afraid, we have to differentiate WARPER_PRESET_FACTORY, IWARPER_PRESET_FACTORY and similar cases, since when detecting errors with .revertedWithCustomErrors only by passing the relevant reference to exact contract/its interface/or library that declares the error/event or whatever.

Comment (alan-parsiq) [reply src/protocol/warper/preset-factory/errors.ts]: I think we can discuss, depending on that I might have couple of "new-style" places where I would rewatch

File: src/protocol/wizards/helpers/general-v1/errors.ts
Comment (alan-parsiq) [inline src/protocol/wizards/helpers/general-v1/errors.ts]: .../helpers/... is not necessary! And most probably it could be src/protocol/wizards/v1/general-wizards-helper/errors.ts

Comment (alan-parsiq) [reply src/protocol/wizards/helpers/general-v1/errors.ts]: GENERAL_WIZARDS_HELPER_V1_ERRORS -> GENERAL_WIZARDS_HELPER_V1_CUSTOM_ERRORS

File: test/builders/warper/warper-deploy-from-preset-builder.ts
Comment (alan-parsiq) [inline test/builders/warper/warper-deploy-from-preset-builder.ts:1]: This one looks fantastic 👍

File: test/builders/wizards-v1/universe/universe-erc721-warper-builder-v1.ts
Comment (alan-parsiq) [inline test/builders/wizards-v1/universe/universe-erc721-warper-builder-v1.ts]: Here

1. My personal outlook, but I would shift from leaving default values in Builders. And the reason, so that everyone who uses Builders would **>intentionally<** assign all values without relying on defaults. As an example check UniverseConstructorV1.anyUniverse
2. I think this time this does not require improvement, since I guss wht can fit really well instead of warperAddress, taxTerms and warperName - the WarperRegistrationConstructorV1. So you can just add the WarperRegistrationConstructorV1 instead of WarperConstructorV1, because the first one already contains the second one :)

File: test/builders/wizards-v1/universe/universe-erc721-warper-builder-v1.ts
Comment (alan-parsiq) [inline test/builders/wizards-v1/universe/universe-erc721-warper-builder-v1.ts]: So I guess here had to be this.warperConstructor.getWarperPresetId, but once you migrate to WarperRegistrationConstructorV1 - you can use WarperRegistrationConstructorV1.prepareWarperRegistrationParams here

File: test/constructors/warper/invalid-warper-constructor.ts
Comment (alan-parsiq) [inline test/constructors/warper/invalid-warper-constructor.ts]: How about we change the name to EmptyDataWarperFromPresetConstructor? Because 

  private presetId: BytesLike = EMPTY_BYTES32_DATA_HEX;
  private initData: BytesLike = EMPTY_BYTES_DATA_HEX;
  
In one context can be pretty valid :) And then you could reuse it in WarperRegistrationConstructorV1, where in prepareWarperRegistrationParams I placed a couple of moments, like "this.warperConstructor ? this.warperConstructor.getWarperPresetId() : EMPTY_BYTES32_DATA_HEX" and we could just replace it with "this.WarperConstructor" and provide EmptyDataWarperConstructor to this Builder in cases, where we need empty data Warper.

Which you an see, can be either

1. wrong behavior;
2. or a sign that Warper Does not have to be created from the Warper Preset.

And, finally, I would propose to change WarperConstructorV1 to WarperFromPresetConstructorV1 and add empty  subclass of Erc721WarperFromPresetConstructorV1 (which would be a sign of ERC721 Warper Preset domain). This might just come handy when narrowing to specific class (since integrating 1155 in IQ is a question of time :) )

File: test/integration/universe/wizards-v1/universe-lifecycle/basic-paths/scenario-2.ts
Comment (alan-parsiq) [inline test/integration/universe/wizards-v1/universe-lifecycle/basic-paths/scenario-2.ts]: taxTerms -> universeWarperTaxTerms, since it reminds about specifics of the Tax Terms that we apply them to the given Universe -> registered Warper level.

File: test/integration/universe/wizards-v1/universe-lifecycle/basic-paths/scenario-2.ts
Comment (alan-parsiq) [inline test/integration/universe/wizards-v1/universe-lifecycle/basic-paths/scenario-2.ts]: Theoretically, you could use .anyUniverse here?

File: test/integration/universe/wizards-v1/universe-lifecycle/basic-paths/scenario-2.ts
Comment (alan-parsiq) [inline test/integration/universe/wizards-v1/universe-lifecycle/basic-paths/scenario-2.ts]: Myself forgot to use ZERO in one other example, could you use here the ZERO constant and also check other place, where the same situation (search for vailabilityPeriodStart: 0)

File: test/integration/universe/wizards-v1/universe-lifecycle/basic-paths/scenario-2.ts
Comment (alan-parsiq) [inline test/integration/universe/wizards-v1/universe-lifecycle/basic-paths/scenario-2.ts]: constant EMPTY_STRING

File: test/integration/universe/wizards-v1/universe-lifecycle/basic-paths/scenario-2.ts
Comment (alan-parsiq) [inline test/integration/universe/wizards-v1/universe-lifecycle/basic-paths/scenario-2.ts]: At the moment when create this UNIVERSE_REGISTRY_CUSTOM_ERRORS, did not think about separating ,but would be great if you could level UNIVERSE_REGISTRY_CUSTOM_ERRORS to wht related to UniverseRegistry and move what related to IUniverseRegistry to IUNIVERSE_REGISTRY_CUSTOM_ERRORS 👍

File: test/integration/universe/wizards-v1/universe-lifecycle/basic-paths/scenario-2.ts
Comment (alan-parsiq) [inline test/integration/universe/wizards-v1/universe-lifecycle/basic-paths/scenario-2.ts]: An idea, to move those bad cases to scenario-3 and differentiate scenario-2 and scenario-3 by first being successful/happy paths and second one failure paths

File: test/integration/universe/wizards-v1/universe-lifecycle/basic-paths/scenario-2.ts
Comment (alan-parsiq) [inline test/integration/universe/wizards-v1/universe-lifecycle/basic-paths/scenario-2.ts]: Same as UNIVERSE_REGISTRY_CUSTOM_ERRORS

File: test/integration/universe/wizards-v1/universe-lifecycle/basic-paths/test-suite.ts
Comment (alan-parsiq) [inline test/integration/universe/wizards-v1/universe-lifecycle/basic-paths/test-suite.ts]: Would be great to stress: Basic Universe with (${PROTOCOL_SUPPORTED_ASSET_CLASSES.ERC721}) Warper from Preset setup.

File: test/validators/tax/tax-terms-registry-validator.ts
Comment (alan-parsiq) [inline test/validators/tax/tax-terms-registry-validator.ts]: Also with TAX_TERMS_REGISTRY_EVENTS  similar to CUSTOM_ERRORS_SITUATION :)

**Review by alan-parsiq — CHANGES_REQUESTED** [review]

File: test/builders/wizards-v1/universe/universe-erc721-warper-builder-v1.ts
Comment (norbertsj) [reply test/builders/wizards-v1/universe/universe-erc721-warper-builder-v1.ts]: I am using `WarperConstructorV1 ` because it is more general and can be used by any constructor which extends `WarperConstructorV1`. In my case, I was using `Erc721ConfigurableWarperConstructor` with this builder.

File: test/builders/wizards-v1/universe/universe-erc721-warper-builder-v1.ts
Comment (norbertsj) [reply test/builders/wizards-v1/universe/universe-erc721-warper-builder-v1.ts]: Yes, and I realized that this will be a generic universe with warper builder. I do not know about `WarperRegistrationConstructorV1 ` - I looked at it, realized I do not want to use it because I think it is meant for a specific purpose. So that is why I created another one.

File: test/integration/universe/wizards-v1/universe-lifecycle/basic-paths/scenario-2.ts
Comment (norbertsj) [reply test/integration/universe/wizards-v1/universe-lifecycle/basic-paths/scenario-2.ts]: Umm, what is the difference? I mean, between UNIVERSE_REGISTRY and IUNIVERSE_REGISTRY? (solidity ultra beginner here).

File: test/integration/universe/wizards-v1/universe-lifecycle/basic-paths/scenario-2.ts
Comment (norbertsj) [reply test/integration/universe/wizards-v1/universe-lifecycle/basic-paths/scenario-2.ts]: Then we probably should do the same for scenario-1?

File: test/builders/wizards-v1/universe/universe-erc721-warper-builder-v1.ts
Comment (alan-parsiq) [reply test/builders/wizards-v1/universe/universe-erc721-warper-builder-v1.ts]: WarperRegistrationConstructorV1 is also general! Take a look at it 👍

File: test/integration/universe/wizards-v1/universe-lifecycle/basic-paths/scenario-2.ts
Comment (alan-parsiq) [reply test/integration/universe/wizards-v1/universe-lifecycle/basic-paths/scenario-2.ts]: One is the actual contract (it is getting physically deployed on Blockchain) and another one is interface of this contract. This is just our convention that UniverseRegistry (and similar) have IUniverseRegistry (and similar) named interface. 

Since for example with custom errors - you must specifically provide typechain interface of contract/interface/library where the custom error is declared. And it just makes sense to separate errors and events by the place of their declaration 👍

File: test/integration/universe/wizards-v1/universe-lifecycle/basic-paths/scenario-2.ts
Comment (alan-parsiq) [reply test/integration/universe/wizards-v1/universe-lifecycle/basic-paths/scenario-2.ts]: I think, we have a flexibility. I am just giving my vote for compact scnarios with minimum viable interfaces. Sometimes it might be an overhead to split to 2 separate scenarios, when you have just 2 test cases. In your piece - you have more and we could improve the readability!

File: test/integration/universe/wizards-v1/universe-lifecycle/basic-paths/scenario-2.ts
Comment (norbertsj) [reply test/integration/universe/wizards-v1/universe-lifecycle/basic-paths/scenario-2.ts]: > would be great if you could level UNIVERSE_REGISTRY_CUSTOM_ERRORS to wht related to UniverseRegistry and move what related to IUniverseRegistry to IUNIVERSE_REGISTRY_CUSTOM_ERRORS

would this mean that we could have both UNIVERSE and IUNIVERSE errors? Not sure if I understand this. Seems to me that UNIVERSE.interface should have everything that is in IUNIVERSE? Or do you mean rename UNIVERSE errors to IUNIVERSE errors?

Comment (alan-parsiq) [reply test/integration/universe/wizards-v1/universe-lifecycle/basic-paths/scenario-2.ts]: When writing contracts, you can declare the errors either in contracts/abstract contracts/interfaces. Unfortunately, I think this is specifics of hardhat matchers that you must provide specifically the place where it is declared.

So if contract implements interface, here it does not mean that declaration is also inside the contract.

So that is why this difference is crucial.


Most of times we will have I<CONTRACT_NAME>_CUSTOM_ERRORS, since I think mostly we were trying to declare them in interfaces. As well as in libraries - which are separate <LIB_BAME>_CUSTOM_ERRORS. Sometimes inside the contract itself, making <CONTRACT_NAME>_CUSTOM_ERRORS

Comment (norbertsj) [reply test/integration/universe/wizards-v1/universe-lifecycle/basic-paths/scenario-2.ts]: Thanks for explanation, now it makes sense :) Will make an update!

**Review by alan-parsiq — APPROVED** [review]
All good, just last very-very minor things:
Move to test/utils/common-test-stuff/dummy-data/warper.ts, the following:
export const COMMON_TEST_WARPER_NAME_1 = TEST_PURPOSE(‘OG Warper’);
WarperRegistrationConstructorV1.warperConstructor -> WarperRegistrationConstructorV1.warperFromPresetConstructor
UniverseWithWarperBuilderV1.warperConstructor -> UniverseWithWarperBuilderV1.warperRegistrationConstructor
And thats it. I am putting you an approve! And once you finish - you can merge!

**Pull request merged**
